Rossen physics

brickman1444 opened the pull request:
Changing the actor to use the Rossen Physics with the lifter and bumper. Should be reasonably usable at the moment.

randomPoison committed:
Janky (But Working!) Rossen Physics
- Player is now configured with a lifter and bumper to do environmental collisions.
- Temporarily gutted ActorPhysics for testing purposes.
- Fixed the editor crash in DebugUtils.Assert().

randomPoison committed:
Working Rosen Physics
- The lifter-bumper combo now works correctly and feels great!
- Added a screenshot button (T right now).

randomPoison committed:
Orient to Movement
- ActorPhysics now orients to the direction of movement, rather than the input direction.
- ActorPhysics now orients the root object (instead of the model) to avoid problems with animations overriding the rotation.
- Fixed bugs in WadeUtils.IsZero() for vectors (wasn't returning correct values).
- Made the WadeUtiles helper methods for vector to be extension methods for convenience.

randomPoison committed:
Better Actor Orientation
- ActorPhysics now factors intended movement in with actual movement, giving a better looking effect.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -12,28 +12,32 @@ public enum ActorStates
 
 public class ActorPhysics : ActorComponent
 {
+	private new Transform transform;
```
Why do this instead of naming a new variable? I get that we don't want to be doing getComponent all the time, but this seems hacky.

brickman1444 replied:
The idea is that the interface is the same but with only one GetComponent() call.

cjacobwade replied:
It's just weird to me that it depends on Awake for transform to work. If anything else tries to refer to this before awake it will get a null ref. I'd prefer something separate like myTransform so the getComponent functionality is still there as well.

randomPoison replied:
Yeah, I've noticed that when doing things like gizmos and editor scripting it causes problems. Unity also has `Reset()` which get's called in the editor, but if that doesn't work for some reason we could do lazy initialization like we're doing `instance` for `Singleton`.

cjacobwade replied:
How would the implementation for lazy initialized this work? Wouldn't you need another variable anyway?

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,52 +47,47 @@ public float groundedMoveSpeed
 
-	protected float groundSlopeSpeedMod = 1f;
-	float groundSlopeSpeedModeDelta = 2.5f;
+	private Vector3 _lastPos;
```
Why are we labeling private variable with underscores? Looks weird and doesn't really give any extra info. If a variable is private, I'll notice when I try to use it.

brickman1444 replied:
Yea David and I have tried to start doing it. It helps when writing a system to quickly visually see what's public and private.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -301,40 +346,6 @@ public void StopClimbing()
 		}
 	}
 
-	protected void CheckGroundSlope()
```
Removing this brings back the problem of sliding up Unity terrain. It's still happens and I think we need a solution before we remove this.

randomPoison replied:
The lifter/bumper combo should take care of this. If the slope is steep enough Then it will collide with the bumper and push the player back from the wall. We can tweak the maximum climbable slope by changing the configuration of the colliders.

cjacobwade replied:
Did you get this feeling right or are we going with invisible walls on every steep slope? We should mention it to Tasha if so.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -362,15 +373,19 @@ public void JumpCheck()
 			isOnGround = true;
 		}
 
-		if ( ( isOnGround || lateJumpTimer < lateJumpTime ) )
```
Bring back late jump before committing please

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -380,77 +395,92 @@ public void JumpCheck()
-					//GainControl();
-				}
-
-				actor.animator.SetBool( "isJumping", false );
```
This is removing the cue for the jump animation so this will need to be added back in.

brickman1444 replied:
I'm not totally sure what's going on here but it looks like it's a case where it is exiting the jump state and reentering grounded. So it's not triggering the jump animation, it's stopping it. Does it still seem necessary?

cjacobwade replied:
There's a big logic chunk that got cut off. We basically need to put logic back in to control animation states somewhere since it's taken out here.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -380,77 +395,92 @@ public void JumpCheck()
+	 *     model will face the direction of movement, rather
+	 *     than the input direction.
+	 */
+	public void OrientSelf()
```
Something in here or in climbing causes a weird bug where the player model rotation is offset and doesn't return to normal.

randomPoison replied:
Yeah, I've got a fix. I changed `ActorPhysics` to rotate the root transform rather than the model's transform, but there were other places in code that were still rotation the model. I cleaned those up and that fixed the problem.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -380,77 +395,92 @@ public void JumpCheck()
-		Gizmos.DrawSphere( transform.position + model.forward * climbCheckDistance, climbCheckRadius );
+		if ( Application.isPlaying )
+		{
+			Gizmos.DrawWireSphere( GetComponent<Transform>().position + bumper.center,
```
Why should this only be when playing? This goes back to why we shouldn't use new private Transform. Calling GetComponent<Transform> looks totally weird.

brickman1444 replied:
changed to if ( transform )

cjacobwade commented on Assets/Scripts/Tags/ClimbableTag.cs:
```diff
@@ -3,16 +3,16 @@
 {
-	[SerializeField] bool xMovement;
-	public bool XMovement
+	[SerializeField] bool _xMovement = true;
```
I like this use case where the private variables accessed through an instance have underscores, but don't think usual private variables need an underscore.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/PlayerActorPhysics.cs:
```diff
@@ -34,7 +32,11 @@ public override void SetupStateMethodMap()
 	void Jumping()
 	{
-		JumpCheck();
+		// TODO should we be doing JumpCheck() while jumping?
```
GetButtons during FixedUpdate sometimes miss button presses. Should change this to set a flag every frame if input is down.

randomPoison commented:
@cjacobwade @brickman1444 I just want to make it clear that the physics refactoring is **not** done yet, that is why there are so many problems and things that have not been implemented.

cjacobwade commented:
Yep! That was clear, I didn't look to much at syntax. Mostly just marking TODOs.

randomPoison commented:
Ah, cool.

randomPoison commented:
Also, btw, it looks here like a bunch of important things were deleted, but they weren't. I had them commented out while working on other stuff and Zac deleted them in his cleanup commit. I'll be rolling back his changes and putting back most of that code.

randomPoison committed:
Ground Collision and Jump Refactoring
- Refactored jump code so that the PlayerActor base class doesn't check input.
- Fixed bug that was causing the player to be rotated wrong after climbing.

randomPoison committed:
Re-Implement Late Jump Delay
- Move the player actor input code to check input in Update() rather than FixedUpdate().
- Added a minimum delay between jumping and landing to prevent issues with immediately colliding with the ground after jumping.

randomPoison commented:
I made a bunch more changes. Jump delay has been added back in and various issues have been fixed. The climbing system still is in need of a cleanup pass in light of the new collisions, but most stuff is working now and platforming feels pretty good. @brickman1444 @cjacobwade Give this another look over if you have a chance before we create the build, otherwise it's ready for the show build at least.

randomPoison committed:
Working Climb From Ground
- You can now start climbing while touching the ground (previously you had to be in the air or it wouldn't work).
- Added a ReadOnly attribute for displaying non-editable information in the inspector.

randomPoison committed:
Climbing Cleanup
- Input for climbing is now checked in PlayerActorPhysics rather than ActorPhysics.
- No longer perform sphere cast for climb check.
- PlayerActorStats no longer checks PlayerActorPhysics for climbing.

randomPoison committed:
Add Lifter/Bumper to Buddy Prefab

randomPoison committed:
Climbing Movement Relative to Climb Volume
- Movement while climbing is now relative to the climb volume.
- Also lots of changes to how movement works are made. It's a lot smoother now, but it's also kind of broken.
- Remove WadeUtils.IsNegative() for vectors because vectors can't have a negative magnitude.
    - Also removed IsPositive() because it's redundant with IsZero() around.

randomPoison committed:
Less Shitty Climb Physics
- Player no longer can climb into the ground or phase into the wall while climbing.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -6,34 +6,39 @@ public enum ActorStates
+	#region States
 	public delegate void ActorStateMethod();
 
+	[ReadOnlyAttribute, SerializeField]
```
Can we name attributes without the word Attribute in them? It would be preferable not to since it can be assumed, but I might recall that being needed...

randomPoison replied:
Yeah, that should actually be just `ReadOnly`, I'm surprised having `ReadOnlyAttribute` didn't cause a build error.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -6,34 +6,39 @@ public enum ActorStates
 	protected Vector3 moveVec = Vector3.zero;
 
-	protected float stoppingSpeed = 0.001f;
+	public float stoppingSpeed = 0.001f;
```
This shouldn't be public. It's referenced in ActorPlayerPhysics but nowhere else.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,113 +48,135 @@ public float groundedMoveSpeed
-	float groundSlopeCapsuleWidth = 0.4f;
-	float groundSlopeCastDist = 0.3f;
+	[Range( 0.0f, 1.0f )]
+	[SerializeField] float lookDirLerp = 0.002f;
```
Can you rename this to something readable in the inspector?

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -6,34 +6,39 @@ public enum ActorStates
 	float stopMoveTimer = 0f;
 
-	[SerializeField] float maxSpeed = 40f;
 	[SerializeField] float _groundedMoveSpeed = 6f;
```
Either all or non of these private, serialized vars should have the underscore.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,113 +48,135 @@ public float groundedMoveSpeed
+	[SerializeField] float jumpCheckRadius = 0.5f;
+
 	[SerializeField] LayerMask jumpLayer = 0;
 	[SerializeField] float minJumpDot = 0.4f;
```
This could probably have a better name. It's clear if dot product is second nature, but totally obtuse otherwise.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,113 +48,135 @@ public float groundedMoveSpeed
 	[SerializeField] float climbCheckRadius = 0.7f;
-	[SerializeField] float surfaceHoldForce = 0.1f;
 
 	[SerializeField] float climbCheckTime = 0.2f;
```
This isn't used anymore in your implementation.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,113 +48,135 @@ public float groundedMoveSpeed
-	[SerializeField] float surfaceHoldForce = 0.1f;
 
 	[SerializeField] float climbCheckTime = 0.2f;
 	float climbCheckTimer = 1f;
```
climbCheckTime/climbCheckTimer aren't used anymore with your implementation.

randomPoison replied:
They're just commented out, there's big performance hit when not using them but it feels way worse when using the timers, so I want to leave this in until I can put together a better solution.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -43,113 +48,135 @@ public float groundedMoveSpeed
-	}
+	ClimbableTag climbTag;
+	Vector3 climbSurfaceNormal = Vector3.zero;
+	Vector3 climbSurfaceRight = Vector3.zero;
```
If we have climbSurface saved can't we just get these when we need them? Would be easy to write your z-in/out-of-wall check to be reusable.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -6,34 +6,39 @@ public enum ActorStates
 
-	protected float stoppingSpeed = 0.001f;
+	public float stoppingSpeed = 0.001f;
 	protected float currStoppingPower = 0.0f;
```
This is no longer being used anywhere.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -182,101 +209,99 @@ public void ComeToStop()
-				if ( !ct.YMovement ) surfaceRelativeInput.y = 0f;
+			Vector3 surfaceRelativeInput =
+			    climbSurfaceRight * ( climbTag.xMovement ? movement.x : 0.0f ) +
+			    Vector3.up * ( climbTag.yMovement ? movement.z : 0.0f );
```
Formatting is weird here, probably should split this up.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -182,101 +209,99 @@ public void ComeToStop()
-			{
-				Debug.LogError("Cannot climb, surface isn't tagged. This is probably a problem in ClimbCheck");
-			}
+			transform.Translate( moveVec * Time.deltaTime, Space.World );
```
This makes me so unhappy, but only because I hate transform.Translate irrationally. I'd totally set velocity here.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -182,101 +209,99 @@ public void ComeToStop()
+		rigidbody.velocity = Vector3.zero;
+		lifter.gameObject.SetActive( false );
+		bumper.gameObject.SetActive( false );
+		climbBumper.gameObject.SetActive( true );
```
Probably should split these into a function like SetClimbingCollidersActive.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -286,13 +311,16 @@ public void StartClimbing( Collider col )
+			rigidbody.useGravity = true;
+			lifter.gameObject.SetActive( true );
+			bumper.gameObject.SetActive( true );
+			climbBumper.gameObject.SetActive( false );
```
Then use SetClimbingCollidersActive(false) here.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -403,54 +409,113 @@ public void JumpCheck()
+		if ( IsInState( ActorStates.Climbing ) )
 		{
-			jumpColCheckTimer += Time.deltaTime;
+			Quaternion.LookRotation( climbSurfaceNormal );
```
Guessing this should be desiredLook = Quaternion.LookRotation( climbSurfaceNormal );

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -403,54 +409,113 @@ public void JumpCheck()
+			Vector3 intendedVelocity = rigidbody.velocity;
+			intendedVelocity.y = 0.0f;
+
+			Vector3 lookVec = Vector3.Lerp( actualVelocity, intendedVelocity, lookDirLerp );
```
lookDirLerp should probably be named something like lookIntentionWeight.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -403,54 +409,113 @@ public void JumpCheck()
-		model.position = transform.position - modelOffset;
+		_lateJump = false;
+	}
+	#endregion
```
Hadn't thought of using Invoke this way. Clever! We should wrap this in a Timer class and have like StartTimer, EndTimer, CheckTimer and CancelTimer with an enum as a timerID.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs:
```diff
@@ -403,54 +409,113 @@ public void JumpCheck()
-		Gizmos.color = Color.green;
-		Gizmos.DrawSphere( transform.position, climbCheckRadius );
-		Gizmos.DrawSphere( transform.position + model.forward * climbCheckDistance, climbCheckRadius );
+		if ( Application.isPlaying )
```
Rip this out before merging.

cjacobwade commented on Assets/Scripts/Actors/ActorComponents/PlayerActorPhysics.cs:
```diff
@@ -96,30 +117,33 @@ void JumpMovement()
 			rigidbody.velocity = moveVec;
 
-			if ( actor.animator )
+			if ( _actor.animator )
```
Other spots have _actor.animator != null, I prefer this shorter way of doing it. Either way pick one and be consistent.

cjacobwade commented on Assets/Scripts/Utilities/DebugUtils.cs:
```diff
@@ -8,7 +8,7 @@ public static void Assert( bool condition, string message = "" )
 			Debug.LogError( message );
-			System.Diagnostics.Debugger.Break();
+			// Debug.Break(); // Uncomment if you want to pause the editor when an assert is hit.
 		}
```
This function should have a bool like BreakOnAssert with default value false. Then Debug.Break only happens if that's true.

cjacobwade commented:
There's a bunch of commented code all over and the underscoring is somewhat inconsistent. I say lets do them on all private variables serialized or not.

randomPoison committed:
Merge branch 'master' into rossen_physics

Conflicts:
	Assets/Environment/SmallMountains.asset
	Assets/Prefabs/Actors/Buddy.prefab
	Assets/Prefabs/Actors/Player.prefab
	Assets/Scenes/Garden.unity
	Assets/Scripts/Actors/ActorComponents/ActorPhysics.cs
	Assets/Scripts/Actors/ActorComponents/PlayerActorStats.cs

randomPoison committed:
Climb Sloped Surfaces
- Climbing orients correctly to sloped surfaces.
- Lean slightly to match sloped surfaces.

randomPoison committed:
Code Review Cleanup
- Miscellaneous cleanup as requested in the code review.
- All private members have underscores.
- No more protected members.

randomPoison committed:
Merge remote-tracking branch 'origin/master' into rossen_physics

Conflicts:
	Assets/Prefabs/Actors/Player.prefab
	Assets/Scenes/Garden.unity
	ProjectSettings/TagManager.asset
